data-prepper-plugins/otel-trace-group-processor/src/main/java/com/amazon/dataprepper/plugins/processor/oteltracegroup/OTelTraceGroupProcessor.java

```diff
@@ -0,0 +1,204 @@
+                optionalStringTraceGroupEntry.ifPresent(entry -> traceIdToTraceGroup.put(entry.getKey(), entry.getValue()));
+            });
+        } catch (Exception e) {
+            // TODO: retry for status code 429 of OpenSearchException?
+            LOG.error("Search request for traceGroup failed for traceIds: {} due to {}", traceIds, e.getMessage());
+        }
+
+        return traceIdToTraceGroup;
+    }
+
+    private SearchRequest createSearchRequest(final Collection<String> traceIds) {
+        final SearchRequest searchRequest = new SearchRequest(OTelTraceGroupProcessorConfig.RAW_INDEX_ALIAS);
+        final SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.query(
+                QueryBuilders.boolQuery()
+                        .must(QueryBuilders.termsQuery(OTelTraceGroupProcessorConfig.TRACE_ID_FIELD, traceIds))
+                        .must(QueryBuilders.termQuery(OTelTraceGroupProcessorConfig.PARENT_SPAN_ID_FIELD, ""))
+        );
+        searchSourceBuilder.docValueField(OTelTraceGroupProcessorConfig.TRACE_ID_FIELD);
+        searchSourceBuilder.docValueField(TraceGroup.TRACE_GROUP_NAME_FIELD);
+        searchSourceBuilder.docValueField(TraceGroup.TRACE_GROUP_END_TIME_FIELD, OTelTraceGroupProcessorConfig.STRICT_DATE_TIME);
+        searchSourceBuilder.docValueField(TraceGroup.TRACE_GROUP_DURATION_IN_NANOS_FIELD);
+        searchSourceBuilder.docValueField(TraceGroup.TRACE_GROUP_STATUS_CODE_FIELD);
+        searchSourceBuilder.fetchSource(false);
+        searchRequest.source(searchSourceBuilder);
+
+        return searchRequest;
+    }
+
+    private Optional<Map.Entry<String, TraceGroup>> fromSearchHitToMapEntry(final SearchHit searchHit) {
+        final DocumentField traceIdDocField = searchHit.field(OTelTraceGroupProcessorConfig.TRACE_ID_FIELD);
+        final DocumentField traceGroupNameDocField = searchHit.field(TraceGroup.TRACE_GROUP_NAME_FIELD);
+        final DocumentField traceGroupEndTimeDocField = searchHit.field(TraceGroup.TRACE_GROUP_END_TIME_FIELD);
+        final DocumentField traceGroupDurationInNanosDocField = searchHit.field(TraceGroup.TRACE_GROUP_DURATION_IN_NANOS_FIELD);
+        final DocumentField traceGroupStatusCodeDocField = searchHit.field(TraceGroup.TRACE_GROUP_STATUS_CODE_FIELD);
+        if (Stream.of(traceIdDocField, traceGroupNameDocField, traceGroupEndTimeDocField, traceGroupDurationInNanosDocField,
+                traceGroupStatusCodeDocField).allMatch(Objects::nonNull)) {
+            final String traceId = traceIdDocField.getValue();
+            final String traceGroupName = traceGroupNameDocField.getValue();
+            // Restore trailing zeros for thousand, e.g. 2020-08-20T05:40:46.0895568Z -> 2020-08-20T05:40:46.089556800Z
+            final String traceGroupEndTime = Instant.parse(traceGroupEndTimeDocField.getValue()).toString();
```

Comment:
Rather than add an inline comment (which is prone to being lost over time), I suggest you make this a function.

Perhaps:

```
/**
 * Restores trailing zeros for thousand, e.g. 2020-08-20T05:40:46.0895568Z -> 2020-08-20T05:40:46.089556800Z
 */ 
private String normalizeDateTime(String dateTimeString) {
  return Instant.parse(dateTimeString).toString();
}
```